design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,300 @@
+- The admin can get metadata specific to the bucket and its object store by executing a `kubectl describe` on a particular ObjectBucket.
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want to leverage the Kubernetes API to create object store buckets. I expect to get back the bucket endpoint information in a Pod-consumable API resource._
+ 
+//TODO update diagram to reflect API changes  
+ 
+![user-actions](./cobc-user.png)
+
+1. The user creates an OBC in their namespace.
+1. The operator detects the new OBC instance.
+    1. The operator checks the `provisioner` field of the OBC's StorageClass.  If it matches, continue to the next step. Else, end the process here.
+    1. The operator gets the object store service using `storageClass.parameters.objectStoreNamespace` + `storageClass.parameters.objectStore`.
+1. The operator creates a CephObjectUser if it does not exist in order to generate the operator's access key.
+1. The operator creates the bucket using the service endpoint and access key.  The bucket name will be a concatenation of the  OBC's `metadata.namespace` and `metadata.name`.
+    1. If the bucket already exists in the store, the operator will instead regenerate the ConfigMap and Secret associated with that bucket.  This allows users the ability to reacquire buckets who's claims have been deleted.
+1. The operator creates an OB containing metadata about the new bucket and sets it as bound to the OBC.
+1. The operator updates the OBC to set it as "bound" to the OB.
+1. The operator creates a ConfigMap in the namespace of the OBC with the endpoint information for that bucket.
+    1. The operator uses the radosgw-admin CLI to generate the user's access key.  
+    1. The key is linked to the bucket.
+1. The operator creates a Secret in the namespace of the OBC with the **user's** access key.
+1. An app Pod may then reference the Secret and the ConfigMap to begin accessing the bucket.
+
+**Use Case: Delete an Object Bucket**
+
+_As a Kubernetes user, I want to delete ObjectBucketClaim instances and cleanup generated API resources._
+
+1. The user deletes the OBC via `kubectl delete objectbucketclaim -n default FOO`.
+1. The OBC is marked for deletion and left in the foreground.
+1. The respective ConfigMap and user Secret are deleted.
+1. The OBC is garbage collected.
+1. The OB phase is changed from "bound" to "lost"
+
+    Note: this does not cause the deletion of the OB.  Since the actual bucket is not deleted, it is reasonable that it should still tracked in Kubernetes.
+
+---
+
+## Looking Forward
+
+- Resource Quotas cannot yet be defined for CRDs.  This limits admin control over how many buckets can be created per object store.  A operator could be made configurable (via ConfigMap) to cap the number of buckets per object store.  A PR exists for enabling quotas on CRDs (https://github.com/kubernetes/kubernetes/pull/72384).  If it is merged into Kubernetes, a ResourceQuota example should be defined for OBCs.
```

Comment:
> Could we not put a property on the storage class that indicates how many buckets to allow?

As I understand it, max-buckets can be set per user.  In this design, we generate a new user/access key for each bucket that is created which is restricted from creating buckets itself.  By relying on ceph to limit bucket creation, we would need to address how we generate these "user" keys.

Secondly, this behavior isn't likely to be implemented in all other object controllers.  By differentiating ceph object from any other object controller, we break down some of the user end abstraction we hope to achieve with the claim api.  This is because the user would now have to know that ceph object sets bucket limits internally where as other controllers may come to rely on quotas.

Resource quotas also provide a uniform way of enforcing bucket limits and informing users of when those limits have be hit, thus abstracting the underlying object store.